junit-jupiter-api/src/main/java/org/junit/jupiter/api/MethodOrderer.java

```diff
@@ -242,4 +249,75 @@ public void orderMethods(MethodOrdererContext context) {
 		}
 	}
 
+	/**
+	 * {@code MethodOrderer} that sorts methods based on their position in the class file,
+	 * which is the order in the source file.
+	 *
+	 * <p>If two methods have the same name, {@code String} representations of
+	 * their formal parameter lists will be used as a fallback for comparing the
+	 * methods.
+	 */
+	class ClassFileMethodOrderer implements MethodOrderer {
+
+		private static final Map<Class<?>, String> CACHE = new HashMap<>();
+
+		@Override
+		public void orderMethods(MethodOrdererContext context) {
+			context.getMethodDescriptors().sort(COMPARATOR);
+		}
+
+		private static final Comparator<MethodDescriptor> COMPARATOR = Comparator //
+				.comparing(ClassFileMethodOrderer::getMethodIndex) //
+				.thenComparing(descriptor -> parameterList(descriptor.getMethod()));
+
+		private static int getMethodIndex(MethodDescriptor methodDescriptor) {
+			Method method = methodDescriptor.getMethod();
+			String pattern = toStringWithLengthPrefix(method);
+			String classFile = CACHE.computeIfAbsent(method.getDeclaringClass(), cls -> new String(read(cls), UTF_8));
```

Comment:
keeping the declaring class in a cache means that it will never be unloaded !

Reply:
Good point!

I guess we'd better store it in the `LruCache` that will be introduced in conjunction with #2012.

Reply:
I think it's better to use a java.lang.ClassValue which is basically a class cache.

Reply:
Interesting! I wasn't aware of `ClassValue`, so thanks for pointing that out.